core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -0,0 +1,438 @@
+  private val transactionTopicPartitionCount = getTransactionTopicPartitionCount
+
+  def enablePidExpiration() {
+    scheduler.startup()
+
+    // TODO: add transaction and pid expiration logic
+  }
+
+  /**
+   * Get the transaction metadata associated with the given transactional id, or null if not found
+   */
+  def getTransaction(transactionalId: String): Option[TransactionMetadata] = {
+    Option(transactionMetadataCache.get(transactionalId))
+  }
+
+  /**
+   * Add a new transaction metadata, or retrieve the metadata if it already exists with the associated transactional id
+   */
+  def addTransaction(transactionalId: String, txnMetadata: TransactionMetadata): TransactionMetadata = {
+    val currentTxnMetadata = transactionMetadataCache.putIfNotExists(transactionalId, txnMetadata)
+    if (currentTxnMetadata != null) {
+      currentTxnMetadata
+    } else {
+      txnMetadata
+    }
+  }
+
+  /**
+   * Validate the given transaction timeout value
+   */
+  def validateTransactionTimeoutMs(txnTimeoutMs: Int): Boolean =
+    txnTimeoutMs <= config.transactionMaxTimeoutMs && txnTimeoutMs > 0
+
+  def transactionTopicConfigs: Properties = {
+    val props = new Properties
+
+    // enforce disabled unclean leader election, no compression types, and compact cleanup policy
+    props.put(LogConfig.UncleanLeaderElectionEnableProp, "false")
+    props.put(LogConfig.CompressionTypeProp, NoCompressionCodec)
+    props.put(LogConfig.CleanupPolicyProp, LogConfig.Compact)
+
+    props.put(LogConfig.MinInSyncReplicasProp, config.transactionLogMinInsyncReplicas.toString)
+    props.put(LogConfig.SegmentBytesProp, config.transactionLogSegmentBytes.toString)
+
+    props
+  }
+
+  def partitionFor(transactionalId: String): Int = Utils.abs(transactionalId.hashCode) % transactionTopicPartitionCount
+
+  def coordinatorEpochFor(transactionId: String): Option[Int] = inLock (partitionLock) {
+    ownedPartitions.get(partitionFor(transactionId))
+  }
+
+  def isCoordinatorFor(transactionalId: String): Boolean = inLock(partitionLock) {
+    val partitionId = partitionFor(transactionalId)
+
+    // partition id should be within the owned list and NOT in the corrupted list
+    ownedPartitions.contains(partitionId)
+  }
+
+  def isCoordinatorLoadingInProgress(transactionalId: String): Boolean = inLock(partitionLock) {
+    val partitionId = partitionFor(transactionalId)
+
+    loadingPartitions.contains(partitionId)
+  }
+
+  /**
+   * Gets the partition count of the transaction log topic from ZooKeeper.
+   * If the topic does not exist, the default partition count is returned.
+   */
+  private def getTransactionTopicPartitionCount: Int = {
+    zkUtils.getTopicPartitionCount(Topic.TransactionStateTopicName).getOrElse(config.transactionLogNumPartitions)
+  }
+
+  private def loadTransactionMetadata(topicPartition: TopicPartition) {
+    def highWaterMark = replicaManager.getHighWatermark(topicPartition).getOrElse(-1L)
```

Comment:
Chatted with @junrao offline. I did this mainly to just mimic group coordinator's loading behavior. But after discussing with him I think it is safer to read up to LEO (we probably need to do the same for GC as well)